triedb/pathdb/iterator_fast.go

```diff
@@ -0,0 +1,385 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pathdb
+
+import (
+	"bytes"
+	"fmt"
+	"slices"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// weightedIterator is an iterator with an assigned weight. It is used to prioritise
+// which account or storage slot is the correct one if multiple iterators find the
+// same one (modified in multiple consecutive blocks).
+type weightedIterator struct {
+	it       Iterator
+	priority int
+}
+
+func (it *weightedIterator) Cmp(other *weightedIterator) int {
+	// Order the iterators primarily by the account hashes
+	hashI := it.it.Hash()
+	hashJ := other.it.Hash()
+
+	switch bytes.Compare(hashI[:], hashJ[:]) {
+	case -1:
+		return -1
+	case 1:
+		return 1
+	}
+	// Same account/storage-slot in multiple layers, split by priority
+	if it.priority < other.priority {
+		return -1
+	}
+	if it.priority > other.priority {
+		return 1
+	}
+	return 0
+}
+
+// fastIterator is a more optimized multi-layer iterator which maintains a
+// direct mapping of all iterators leading down to the bottom layer.
+type fastIterator struct {
+	db   *Database   // Database to reinitialize stale sub-iterators with
+	root common.Hash // Root hash to reinitialize stale sub-iterators through
+
+	curAccount []byte
+	curSlot    []byte
+
+	iterators []*weightedIterator
+	initiated bool
+	account   bool
+	fail      error
+}
+
+// newFastIterator creates a new hierarchical account or storage iterator with one
+// element per diff layer. The returned combo iterator can be used to walk over
+// the entire layer stack simultaneously.
+func newFastIterator(db *Database, root common.Hash, account common.Hash, seek common.Hash, accountIterator bool) (*fastIterator, error) {
+	current := db.tree.get(root)
+	if current == nil {
+		return nil, fmt.Errorf("unknown layer: %x", root)
+	}
+	fi := &fastIterator{
+		db:      db,
+		root:    root,
+		account: accountIterator,
+	}
+	for depth := 0; current != nil; depth++ {
+		if accountIterator {
+			switch dl := current.(type) {
+			case *diskLayer:
+				fi.iterators = append(fi.iterators, &weightedIterator{
+					// The state set in the disk layer is mutable, and the entire state becomes stale
+					// if a diff layer above is merged into it. Therefore, staleness must be checked,
+					// and the storage slot should be retrieved with read lock protection.
+					it: newDiffAccountIterator(seek, dl.buffer.states, func(hash common.Hash) ([]byte, error) {
+						dl.lock.RLock()
+						defer dl.lock.RUnlock()
+
+						if dl.stale {
+							return nil, errSnapshotStale
+						}
+						return dl.buffer.states.mustAccount(hash)
+					}),
+					priority: depth,
+				})
+				fi.iterators = append(fi.iterators, &weightedIterator{
+					it:       newDiskAccountIterator(dl.db.diskdb, seek),
+					priority: depth + 1,
+				})
+			case *diffLayer:
+				// The state set in diff layer is immutable and will never be stale,
```

Comment:
Do we never merge into layers?

Reply:
For diff layers in the pathdb, yes they are immutable and we never modify the content of them.

The only mutable container is the buffer in the disk layer, which will accept the state set from the diff layers on top.

Reply:
> The only mutable container is the buffer in the disk layer, which will accept the state set from the diff layers on top.

That's nice. We could do the same in the snap layers. Instead of having `n` mutable layers, where only the bottom-most difflayer is the one that is _actually_ mutable, we could treat them differently.